runtime/src/identity.rs

```diff
@@ -778,10 +839,115 @@ decl_event!(
 
         /// DID queried
         DidQuery(Key, IdentityId),
+
+        /// New authorization added (auth_id, from, to, authorization_data, expiry)
+        NewAuthorization(
+            u64,
+            IdentityId,
+            IdentityId,
+            AuthorizationData,
+            Option<Moment>
+        ),
+
+        /// Authorization revoked or consumed. (auth_id, authorized_identity)
+        AuthorizationRemoved(u64, IdentityId),
     }
 );
 
 impl<T: Trait> Module<T> {
+    pub fn add_auth(
+        from_did: IdentityId,
+        target_did: IdentityId,
+        authorization_data: AuthorizationData,
+        expiry: Option<T::Moment>,
+    ) {
+        let new_nonce = Self::multi_purpose_nonce() + 1u64;
+        // Even if this transaction fails, nonce should be increased for added unpredictability
```

Comment:
can the transaction actually fail after this?

Reply:
Nope, I'll remove the comment.